server/src/main/java/org/elasticsearch/transport/RemoteClusterConnection.java

```diff
@@ -110,24 +113,41 @@ void collectNodes(ActionListener<Function<String, DiscoveryNode>> listener) {
             try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
                 // we stash any context here since this is an internal execution and should not leak any existing context information
                 threadContext.markAsSystemContext();
-
-                final ClusterStateRequest request = new ClusterStateRequest();
-                request.clear();
-                request.nodes(true);
-                request.local(true); // run this on the node that gets the request it's as good as any other
                 Transport.Connection connection = remoteConnectionManager.getAnyRemoteConnection();
-                transportService.sendRequest(
-                    connection,
-                    ClusterStateAction.NAME,
-                    request,
-                    TransportRequestOptions.EMPTY,
-                    new ActionListenerResponseHandler<>(
-                        contextPreservingActionListener.map(response -> response.getState().nodes()::get),
-                        ClusterStateResponse::new
-                    )
-                );
+
+                if (RemoteClusterPortSettings.REMOTE_CLUSTER_PROFILE.equals(connectionStrategy.getTransportProfile())) {
+                    transportService.sendRequest(
+                        connection,
+                        RemoteClusterNodesAction.NAME,
+                        RemoteClusterNodesAction.Request.INSTANCE,
+                        TransportRequestOptions.EMPTY,
+                        new ActionListenerResponseHandler<>(contextPreservingActionListener.map(response -> {
+                            final Map<String, DiscoveryNode> nodeLookup = response.getNodes()
+                                .stream()
+                                .collect(Collectors.toUnmodifiableMap(DiscoveryNode::getId, Function.identity()));
+                            return nodeLookup::get;
+                        }), RemoteClusterNodesAction.Response::new)
+                    );
+                } else {
+                    final ClusterStateRequest request = new ClusterStateRequest();
+                    request.clear();
+                    request.nodes(true);
+                    request.local(true); // run this on the node that gets the request it's as good as any other
+
+                    transportService.sendRequest(
+                        connection,
+                        ClusterStateAction.NAME,
+                        request,
+                        TransportRequestOptions.EMPTY,
+                        new ActionListenerResponseHandler<>(
+                            contextPreservingActionListener.map(response -> response.getState().nodes()::get),
+                            ClusterStateResponse::new
```

Comment:
This clusterState request is maintained so that the new model does not affect the legacy model. But it is totally possible to also have the legacy model use the new RemoteClusterNodes action which should help simplify the code on the caller side.

Reply:
After some more thought, it is not an easy task to replace the clusterState call for the legacy model mainly because the other side may not understand the new action. So I think we should just keep it as is.